src/controller/eme-controller.js

```diff
@@ -342,16 +349,28 @@ class EMEController extends EventHandler {
   }
 
   _onMediaEncrypted (event) {
-    const initDataType = event.initDataType;
-    const initData = event.initData;
-    logger.log(`Media is encrypted using "${initDataType}" init data type`);
+    logger.log(`Media is encrypted using "${event.initDataType}" init data type`);
+
+    if (!this._mediaKeysPromise) {
+      logger.error('Fatal: Media is encrypted but no CDM access or no keys have been requested');
+      this.hls.trigger(Event.ERROR, {
+        type: ErrorTypes.KEY_SYSTEM_ERROR,
+        details: ErrorDetails.KEY_SYSTEM_NO_KEYS,
+        fatal: true
+      });
+      return;
+    }
 
-    this._isMediaEncrypted = true;
-    this._mediaEncryptionInitDataType = initDataType;
-    this._mediaEncryptionInitData = initData;
+    const finallySetKeyAndStartSession = (mediaKeys) => {
+      if (!this._media) {
+        return;
+      }
+      this._attemptSetMediaKeys(mediaKeys);
+      this._generateRequestWithPreferredKeySession(event.initDataType, event.initData);
+    };
 
-    this._attemptSetMediaKeys();
-    this._generateRequestWithPreferredKeySession();
+    // Could use `Promise.finally` but some Promise polyfills are missing it
+    this._mediaKeysPromise.then(finallySetKeyAndStartSession).catch(finallySetKeyAndStartSession);
```

Comment:
this is the main change in https://github.com/video-dev/hls.js to resolve race condition.

Reply:
it will store `media keys creation` into `this._mediaKeysPromise` and making sure `set media keys` is occurred after that.

Reply:
this approach is much better.